File: Chapter-7/shift.py
###########################################################################
### Chapter 7 - Evaluating Fielders                                     ###
# Mathletics: How Gamblers, Managers, and Sports Enthusiasts              #
# Use Mathematics in Baseball, Basketball, and Football                   #
###########################################################################

import pandas as pd
import numpy as np
from scipy import optimize, optimize.LinearConstraint

df = pd.read_csv("shift.csv")

df['changeDistance'] = [(df['Base'].iloc[i] - 1)*90 + df['distance'].iloc[i] for i in range(len(df))]

## return the negative of the objective function since we want to maximize but the built-in optimization functions are minimizers
def obj(x):
	ffielded = np.array([1 if min([abs(x[k]-df['changeDistance'].iloc[i]) for k in range(4)])<=5 else 0 for i in range(len(df))]).mean()
	return -ffielded

## since our objective function is not smooth we are using differential_evolution
res = optimize.differential_evolution(obj, bounds = [(0,8),(0,180),(0,180),(0,180)])

File: Chapter-7/shift.R
###########################################################################
### Chapter 7 - Evaluating Fielders                                     ###
# Mathletics: How Gamblers, Managers, and Sports Enthusiasts              #
# Use Mathematics in Baseball, Basketball, and Football                   #
###########################################################################

library(readr)
library(dplyr)

df <- read_csv("shift.csv")

df <- df %>%
  mutate(changeDistance = (Base - 1) * 90 + distance)

# Define the objective function
# Return the negative of the objective function since we want to maximize,
# but the built-in optimization functions in R are minimizers
obj <- function(x) {
  ffielded <- mean(ifelse(sapply(df$changeDistance, function(k) min(abs(x - k)) <= 5), 1, 0))
  return(-ffielded)
}

# Use the optim function for optimization
res <- optim(par = c(0, 0, 0, 0), fn = obj, method = "Nelder-Mead")

cat("The optimal positions are:\n")
cat(paste("\t", round(res$par[1], 2), "\n"))
cat(paste("\t", round(res$par[2], 2), "\n"))
cat(paste("\t", round(res$par[3], 2), "\n"))
cat(paste("\t", round(res$par[4], 2), "\n"))
cat("and the fraction fielded is expected to be:", round(abs(obj(res$par)), 2), "\n")

